Use faulthandler to dump traceback of timed out processes in unit tests.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54818 Use faulthandler to dump traceback of timed out processes in unit tests.**

Several flaky tests fail due to some sort of timeout and it isn't
clear from the error message in CI where exactly each process is stuck. In this
PR, I've added mechanism to dump the entire python traceback of all python
threads when we encounter a timeout.

Example traceback:

```
Process 3 timed out with traceback:
Current thread 0x00007ff3363ff700 (most recent call first):
  File "torch/testing/_internal/common_distributed.py", line 373 in _event_listener
  File "threading.py", line 870 in run
  File "threading.py", line 932 in _bootstrap_inner
  File "threading.py", line 890 in _bootstrap

Thread 0x00007ff406132180 (most recent call first):
  File "torch/distributed/distributed_c10d.py", line 2477 in barrier
  File "torch/testing/_internal/distributed/rpc/rpc_test.py", line 838 in test_reinit
  File "torch/testing/_internal/dist_utils.py", line 90 in new_test_method
  File "torch/testing/_internal/common_distributed.py", line 292 in wrapper
  File "torch/testing/_internal/common_distributed.py", line 409 in run_test
  File "torch/testing/_internal/common_distributed.py", line 393 in _run
  File "multiprocessing/process.py", line 108 in run
  File "multiprocessing/process.py", line 315 in _bootstrap
  File "multiprocessing/popen_fork.py", line 75 in _launch
  File "multiprocessing/popen_fork.py", line 19 in __init__
  File "multiprocessing/context.py", line 277 in _Popen
  File "multiprocessing/process.py", line 121 in start
```

Differential Revision: [D27378764](https://our.internmc.facebook.com/intern/diff/D27378764/)

pritamdamania committed:
Use faulthandler to dump traceback of timed out processes in unit tests.

Several flaky tests fail due to some sort of timeout and it isn't
clear from the error message in CI where exactly each process is stuck. In this
PR, I've added mechanism to dump the entire python traceback of all python
threads when we encounter a timeout.

Example traceback:

```
Process 3 timed out with traceback:
Current thread 0x00007ff3363ff700 (most recent call first):
  File "torch/testing/_internal/common_distributed.py", line 373 in _event_listener
  File "threading.py", line 870 in run
  File "threading.py", line 932 in _bootstrap_inner
  File "threading.py", line 890 in _bootstrap

Thread 0x00007ff406132180 (most recent call first):
  File "torch/distributed/distributed_c10d.py", line 2477 in barrier
  File "torch/testing/_internal/distributed/rpc/rpc_test.py", line 838 in test_reinit
  File "torch/testing/_internal/dist_utils.py", line 90 in new_test_method
  File "torch/testing/_internal/common_distributed.py", line 292 in wrapper
  File "torch/testing/_internal/common_distributed.py", line 409 in run_test
  File "torch/testing/_internal/common_distributed.py", line 393 in _run
  File "multiprocessing/process.py", line 108 in run
  File "multiprocessing/process.py", line 315 in _bootstrap
  File "multiprocessing/popen_fork.py", line 75 in _launch
  File "multiprocessing/popen_fork.py", line 19 in __init__
  File "multiprocessing/context.py", line 277 in _Popen
  File "multiprocessing/process.py", line 121 in start
```

Differential Revision: [D27378764](https://our.internmc.facebook.com/intern/diff/D27378764/)

[ghstack-poisoned]

rohan-varma commented on torch/testing/_internal/common_distributed.py:
```diff
@@ -394,6 +418,18 @@ def run_test(self, test_name, pipe):
+        for i, process in enumerate(self.processes):
+            if process.exitcode is None:
+                pipe = self.pid_to_pipe[process.pid]
+                pipe.send(MultiProcessTestCase.Event.GET_TRACEBACK)
```
nit: send out all the requests before waiting for each proc so they can be processed in parallel by the subprocesses might be better? Might be overkill for this though

rohan-varma commented on torch/testing/_internal/common_distributed.py:
```diff
@@ -359,9 +360,32 @@ def _spawn_processes(self):
+        threading.Thread(
+            target=MultiProcessTestCase._event_listener,
+            args=(pipe,),
+            daemon=True).start()
```
How is this thread destroyed at the end? Does it need to be?

pritamdamania87 replied:
I used `daemon=True` so we don't have to worry about that.

rohan-varma commented on torch/testing/_internal/common_distributed.py:
```diff
@@ -359,9 +360,32 @@ def _spawn_processes(self):
+                    # Flush buffers and seek to read from the beginning
+                    tmp_file.flush()
+                    tmp_file.seek(0)
+                    pipe.send(tmp_file.read())
```
Can't we exit after this send, since the subprocess shouldn't get multiple dump traceback requests? If so, we shouldn't need the `while True` either right, since `recv()` will block until we get a message?

pritamdamania87 replied:
I just kept this pattern so this event_listener can get extended in the future with maybe other things we want it to do.

rohan-varma reviewed (approved):
Thanks, this is great! It would be awesome to also be able to get C++ stacktraces, something like the equivalent of gdb's `thread apply-all bt`, but I'm guessing that would be quite tough to figure out?

pritamdamania87 commented:
> Thanks, this is great! It would be awesome to also be able to get C++ stacktraces, something like the equivalent of gdb's thread apply-all bt, but I'm guessing that would be quite tough to figure out?

Yes, I plan to look into that as well. I think it might be tricky but is probably possible. For example, I found this library that does what we want: https://github.com/thoughtspot/threadstacks, so I'll look into the implementation to see what is a good way to do this.

pritamdamania committed:
Update on "Use faulthandler to dump traceback of timed out processes in unit tests."


Several flaky tests fail due to some sort of timeout and it isn't
clear from the error message in CI where exactly each process is stuck. In this
PR, I've added mechanism to dump the entire python traceback of all python
threads when we encounter a timeout.

Example traceback:

```
Process 3 timed out with traceback:
Current thread 0x00007ff3363ff700 (most recent call first):
  File "torch/testing/_internal/common_distributed.py", line 373 in _event_listener
  File "threading.py", line 870 in run
  File "threading.py", line 932 in _bootstrap_inner
  File "threading.py", line 890 in _bootstrap

Thread 0x00007ff406132180 (most recent call first):
  File "torch/distributed/distributed_c10d.py", line 2477 in barrier
  File "torch/testing/_internal/distributed/rpc/rpc_test.py", line 838 in test_reinit
  File "torch/testing/_internal/dist_utils.py", line 90 in new_test_method
  File "torch/testing/_internal/common_distributed.py", line 292 in wrapper
  File "torch/testing/_internal/common_distributed.py", line 409 in run_test
  File "torch/testing/_internal/common_distributed.py", line 393 in _run
  File "multiprocessing/process.py", line 108 in run
  File "multiprocessing/process.py", line 315 in _bootstrap
  File "multiprocessing/popen_fork.py", line 75 in _launch
  File "multiprocessing/popen_fork.py", line 19 in __init__
  File "multiprocessing/context.py", line 277 in _Popen
  File "multiprocessing/process.py", line 121 in start
```

Differential Revision: [D27378764](https://our.internmc.facebook.com/intern/diff/D27378764/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Use faulthandler to dump traceback of timed out processes in unit tests."


Several flaky tests fail due to some sort of timeout and it isn't
clear from the error message in CI where exactly each process is stuck. In this
PR, I've added mechanism to dump the entire python traceback of all python
threads when we encounter a timeout.

Example traceback:

```
Process 3 timed out with traceback:
Current thread 0x00007ff3363ff700 (most recent call first):
  File "torch/testing/_internal/common_distributed.py", line 373 in _event_listener
  File "threading.py", line 870 in run
  File "threading.py", line 932 in _bootstrap_inner
  File "threading.py", line 890 in _bootstrap

Thread 0x00007ff406132180 (most recent call first):
  File "torch/distributed/distributed_c10d.py", line 2477 in barrier
  File "torch/testing/_internal/distributed/rpc/rpc_test.py", line 838 in test_reinit
  File "torch/testing/_internal/dist_utils.py", line 90 in new_test_method
  File "torch/testing/_internal/common_distributed.py", line 292 in wrapper
  File "torch/testing/_internal/common_distributed.py", line 409 in run_test
  File "torch/testing/_internal/common_distributed.py", line 393 in _run
  File "multiprocessing/process.py", line 108 in run
  File "multiprocessing/process.py", line 315 in _bootstrap
  File "multiprocessing/popen_fork.py", line 75 in _launch
  File "multiprocessing/popen_fork.py", line 19 in __init__
  File "multiprocessing/context.py", line 277 in _Popen
  File "multiprocessing/process.py", line 121 in start
```

Differential Revision: [D27378764](https://our.internmc.facebook.com/intern/diff/D27378764/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Use faulthandler to dump traceback of timed out processes in unit tests."


Several flaky tests fail due to some sort of timeout and it isn't
clear from the error message in CI where exactly each process is stuck. In this
PR, I've added mechanism to dump the entire python traceback of all python
threads when we encounter a timeout.

Example traceback:

```
Process 3 timed out with traceback:
Current thread 0x00007ff3363ff700 (most recent call first):
  File "torch/testing/_internal/common_distributed.py", line 373 in _event_listener
  File "threading.py", line 870 in run
  File "threading.py", line 932 in _bootstrap_inner
  File "threading.py", line 890 in _bootstrap

Thread 0x00007ff406132180 (most recent call first):
  File "torch/distributed/distributed_c10d.py", line 2477 in barrier
  File "torch/testing/_internal/distributed/rpc/rpc_test.py", line 838 in test_reinit
  File "torch/testing/_internal/dist_utils.py", line 90 in new_test_method
  File "torch/testing/_internal/common_distributed.py", line 292 in wrapper
  File "torch/testing/_internal/common_distributed.py", line 409 in run_test
  File "torch/testing/_internal/common_distributed.py", line 393 in _run
  File "multiprocessing/process.py", line 108 in run
  File "multiprocessing/process.py", line 315 in _bootstrap
  File "multiprocessing/popen_fork.py", line 75 in _launch
  File "multiprocessing/popen_fork.py", line 19 in __init__
  File "multiprocessing/context.py", line 277 in _Popen
  File "multiprocessing/process.py", line 121 in start
```

Differential Revision: [D27378764](https://our.internmc.facebook.com/intern/diff/D27378764/)

[ghstack-poisoned]
